### ISSUE TITLE
Confirm pending user without deep-linking?

### ISSUE THREAD

**Issue by devsales** [realm/realm-js#5872]
Is there a way to confirm a pending user in App Services without using deep-linking or universal links?

It seems to me that this is restricting the user - if the confirmation email is not opened on the phone (e.g. on a laptop), then the link from the email would not lead back to the app, but to the browser. This means that the `confirmUser` function will not be called: `await app.emailPasswordAuth.confirmUser({ token, tokenId });`

The same goes for the password reset process…

Am I missing something?

I know the pending user can be confirmed via the Admin API but that does not seem like the correct way to do that. And there is no endpoint for resetting the password...

**Comment by elle-j** [comment]
Hi @devsales, like you're saying, if the user clicks on the link in the confirmation email from e.g. a laptop, the link should go to an external web app where you can handle the confirmation (the same goes for password reset). (Also see [our docs](https://www.mongodb.com/docs/atlas/app-services/authentication/email-password/#user-confirmation) that explains this in more detail.)

**Alternative - Custom Function:**
However, you can also use a [custom function](https://www.mongodb.com/docs/atlas/app-services/authentication/email-password/#run-a-confirmation-function) instead of a confirmation email. The function will receive the username (email), token, and tokenId, which you can then send to the user in a way you prefer. You'd then return `{ status : "pending" }` from that function when the token and tokenId have been sent to the user.

**Alternative - Auto Confirmation (not recommended for production):**
When developing, you can also configure the app to [immediately confirm users](https://www.mongodb.com/docs/atlas/app-services/authentication/email-password/#automatically-confirm-users) (this does not validate the email provided), but this may not be suitable for production.

I hope that this gives you a better understanding of the options available, let me know if you have more questions or concerns!

**Comment by devsales** [comment]
Hi @elle-j, thanks for your answer! I have read through docs a couple of times but I still don't see a way to confirm the user **without using deep-linking or universal links** to redirect the user back to the app after clicking on the confirmation link.

My understanding from the docs is that in order to confirm the user, I have to make this call - `await app.emailPasswordAuth.confirmUser({ token, tokenId });`. However this can be done only in the app itself, right? So if I don't have deep-linking to redirect the user back to the app, the browser would open, which **cannot** make that call. 

Or am I missing something? Can I use the Node.JS SDK to make that call on the server?

**Comment by elle-j** [comment]
@devsales, yes you would need to set up [deep-linking](https://developer.android.com/training/app-links/deep-linking.html) and/or [universal links](https://developer.apple.com/library/archive/documentation/General/Conceptual/AppSearch/UniversalLinks.html) for the redirect functionality. For calling the `confirmUser()` method, this can [also be done](https://www.mongodb.com/docs/realm/sdk/node/users/manage-email-password-users/#confirm-a-new-user-s-email-address) on a Node.js server via our Node.js SDK.

You would then import `realm` and get access to the App instance via `new Realm.App({ id: 'YOUR_ID' })` (similar to how you'd do it in React Native).

**Comment by elle-j** [comment]
I wouldn't say that repeatedly refreshing custom data is ideal; alternatively, you could try the [MongoDBCollection.watch() API](https://www.mongodb.com/docs/realm/web/mongodb/#watch-for-changes) (this links to our Web SDK docs, but should be the same API). This will receive change streams whenever there are changes to a specific Atlas collection (e.g. the collection you assign for your custom user data). Example:

```js
const collection = user
  .mongoClient('service name')
  .db('database name')
  .collection('custom user data collection name');

for await (const change of collection.watch()) {
  let breakAsyncIterator = false; // Used to signal when to exit the async iterator
  switch (change.operationType) {
    case "insert": {
      const { documentKey, fullDocument } = change;

      // Perform you own logic..

      breakAsyncIterator = true;
      break;
    }
  }
  if (breakAsyncIterator) {
    // Exit async iterator
    break;
  }
}
```

Note that polyfills are required in React Native for `watch()` to work:
* Polyfills for `fetch` and `ReadableStream`: https://www.npmjs.com/package/react-native-polyfill-globals
* Babel plugin enabling async generator syntax: https://npmjs.com/package/@babel/plugin-proposal-async-generator-functions

If you need to add something else to your custom user data when the user is created, you could set up an [Authentication Trigger](https://www.mongodb.com/docs/atlas/app-services/triggers/authentication-triggers/#std-label-authentication-triggers) (with action type `CREATE`) via the App Services UI and run a [custom function](https://www.mongodb.com/docs/atlas/app-services/users/custom-metadata/#user-creation-function).
